### PR TITLE
Convert `<div class="button">` to `<button class="button">`

### PR DESCRIPTION
**Pull request by delvh**
This improves a lot of accessibility shortcomings.
Every possible instance of `<div class="button">` matching the command
`ag '<[^ab].*?class=.*?[" ]button[ "]' templates/ | grep -v 'dropdown'` has been converted when possible.
divs with the `dropdown` class and their children were omitted as
1. more analysis must be conducted whether the dropdowns still work as intended when they are a `button` instead of a `div`.
2. most dropdowns have `div`s as children. The HTML standard disallows `div`s inside `button`s.
3. When a dropdown child that's part of the displayed text content is converted to a `button`, the dropdown can be focused twice 

Further changes include that all "gitea-managed" buttons with JS code received an  `e.preventDefault()` so that they don't accidentally submit an underlying form, which would execute instead of cancel the action.
Lastly, some minor issues were fixed as well during the refactoring.

## Future improvements

As mentioned in https://github.com/go-gitea/gitea/pull/23337#discussion_r1127277391, `<a>`s without `href` attribute are not focusable.
They should later on be converted to `<button>`s.

### CODE REVIEW

File: templates/repo/activity.tmpl [hunk 1]
```diff
@@ -6,12 +6,10 @@
 			<div class="ui right">
 				<!-- Period -->
 				<div class="ui floating dropdown jump filter">
-					<div class="ui basic compact button">
-						<span class="text">
-							{{.locale.Tr "repo.activity.period.filter_label"}} <strong>{{.PeriodText}}</strong>
-							{{svg "octicon-triangle-down" 14 "dropdown icon"}}
-						</span>
-					</div>
+					<span class="ui basic compact button text">
```
Comment (wxiaoguang) [inline templates/repo/activity.tmpl]: span or button?

File: templates/admin/notice.tmpl [hunk 2]
```diff
@@ -23,7 +23,7 @@
 						<tr>
 							<td class="collapsing">
 								<div class="ui fitted checkbox" data-id="{{.ID}}">
-									<input type="checkbox"> <label></label>
+									<input type="checkbox">
```
Comment (wxiaoguang) [inline templates/admin/notice.tmpl:26]: If I remember correctly, it's done by purpose. (I haven't checked it)

And it's an unrelated change (this PR is improving buttons), so I think it's better to keep it as it was.

File: templates/admin/user/edit.tmpl [hunk 3]
```diff
@@ -151,7 +151,7 @@
 
 				<div class="field">
 					<button class="ui green button">{{.locale.Tr "admin.users.update_profile"}}</button>
-					<div class="ui red button show-modal" data-modal="#delete-user-modal">{{.locale.Tr "admin.users.delete_account"}}</div>
+					<button type="button" class="ui red button show-modal" data-modal="#delete-user-modal">{{.locale.Tr "admin.users.delete_account"}}</button>
```
Comment (wxiaoguang) [inline templates/admin/user/edit.tmpl]: The above "admin.auths.delete" button doesn't use the `type=button`, while other JS buttons need the `type=button`. It seems inconsistent.

The purpose of `type=button` is to prevent from submitting the form, could such "preventing" be done by JS? Then no need to write a lot of `type=button` in code.

File: templates/base/delete_modal_actions.tmpl [hunk 4]
```diff
@@ -1,10 +1,10 @@
 <div class="actions">
-	<div class="ui red basic inverted cancel button">
+	<button type="button"{{/* To prevent accidental submission when clicking on no */}} class="ui red basic inverted cancel button">
```
Comment (wxiaoguang) [inline templates/base/delete_modal_actions.tmpl]: Is it handled by JS?  If yes, then no need "type=button"?

File: templates/admin/notice.tmpl [hunk 5]
```diff
@@ -39,13 +39,11 @@
...
-										<form method="post" action="{{AppSubUrl}}/admin/notices/empty">
-											{{.CsrfTokenHtml}}
-											<button type="submit" class="ui red small button">{{.locale.Tr "admin.notices.delete_all"}}</button>
-										</form>
-									</div>
-									<div class="ui floating upward dropdown small button">
+									<form class="ui right" method="post" action="{{AppSubUrl}}/admin/notices/empty">
+										{{.CsrfTokenHtml}}
+										<button type="submit" class="ui red small button">{{.locale.Tr "admin.notices.delete_all"}}</button>
+									</form>
```
Comment (wxiaoguang) [inline templates/admin/notice.tmpl:45]: Maybe  we can keep old code as much as possible to avoid unrelated changes.

File: templates/admin/user/edit.tmpl [hunk 6]
```diff
@@ -189,7 +189,7 @@
 
 				<div class="field">
 					<button class="ui green button">{{$.locale.Tr "settings.update_avatar"}}</button>
-					<a class="ui red button delete-post" data-request-url="{{.Link}}/avatar/delete" data-done-url="{{.Link}}">{{$.locale.Tr "settings.delete_current_avatar"}}</a>
+					<a class="ui red button delete-post" data-request-url="{{.Link}}/avatar/delete" data-done-url="{{.Link}}">{{$.locale.Tr "settings.delete_current_avatar"}}</a>{{/* TODO: Why is this link not focusable? */}}
```
Comment (wxiaoguang) [inline templates/admin/user/edit.tmpl]: Because it doesn't have a `href`.

File: templates/repo/issue/view_content/update_branch_by_merge.tmpl [hunk 7]
```diff
@@ -19,8 +19,8 @@
 						<div class="ui dropdown icon button no-text">
 							{{svg "octicon-triangle-down" 14 "dropdown icon"}}
 							<div class="menu">
-								<div class="item active selected" data-do="{{.Link}}/update">{{$.locale.Tr "repo.pulls.update_branch"}}</div>
-								<div class="item" data-do="{{.Link}}/update?style=rebase">{{$.locale.Tr "repo.pulls.update_branch_rebase"}}</div>
+								<button class="item active selected" data-do="{{.Link}}/update">{{$.locale.Tr "repo.pulls.update_branch"}}</button>
+								<button class="item" data-do="{{.Link}}/update?style=rebase">{{$.locale.Tr "repo.pulls.update_branch_rebase"}}</button>
```
Comment (wxiaoguang) [inline templates/repo/issue/view_content/update_branch_by_merge.tmpl]: It could be `<a>` instead of `<button>`. Then `aria.js` could click them.

---

~~Or make `aria.js` also support to click `button`, like `a`~~ -- but I do not think there is a reason that here must use button, it doesn't look like a button and there is no `ui button` class.

File: templates/repo/activity.tmpl (hunk 1 above)
Comment (delvh) [reply templates/repo/activity.tmpl]: `span`.
I've tested it, and noticed that it can otherwise be focused twice.
Will add a comment.

File: templates/admin/notice.tmpl (hunk 2 above)
Comment (delvh) [reply templates/admin/notice.tmpl:26]: Well, I've tested it: The result is exactly the same.
Either way, at the end the label is generated with the `before` marker and assigned to this checkbox.
That's why I removed it here.

File: templates/admin/notice.tmpl (hunk 5 above)
Comment (delvh) [reply templates/admin/notice.tmpl:45]: The problem I can see with such small changes:
How will they ever be fixed then?
If we always exclude them from PRs, then the only way to get rid of them is by posting its own PR to fix it.
The only PRs I've seen for these things are complete refactors, or bug fixes.
I haven't seen PRs fixing one single disturbance.
That means these small and unnecessary disturbances will stay forever, even if they can be easily fixed.

So yes, I can revert it, if we are fine with keeping these disturbances (in this case un unneeded `div`).

File: templates/admin/user/edit.tmpl (hunk 3 above)
Comment (delvh) [reply templates/admin/user/edit.tmpl]: Oh yeah, looks like I saw this common pattern of `two buttons behind each other means accept and cancel button`, and thus added it here.
I can remove it again.

File: templates/base/delete_modal_actions.tmpl (hunk 4 above)
Comment (delvh) [reply templates/base/delete_modal_actions.tmpl]: At the moment, it isn't.
An easy workaround we could do is `document.querySelectorAll('.cancel').forEach(cancelButton => cancelButton.setAttribute('type', 'button'))`.
That should always work, but it would set it dynamically instead of hardcoded.
It depends on what we want:
- A site that works (mostly) correctly even when JS is disabled
- A site with as minimal code as possible

Both approaches have advantages and disadvantages.

File: templates/repo/issue/view_content/update_branch_by_merge.tmpl (hunk 7 above)
Comment (delvh) [reply templates/repo/issue/view_content/update_branch_by_merge.tmpl]: I am also asking myself now how I came to the conclusion to convert these `div`s?
I think I'll just revert those.

File: templates/admin/user/edit.tmpl (hunk 6 above)
Comment (delvh) [reply templates/admin/user/edit.tmpl]: Ah, sounds like a problem for another PR.
I've just had a look, this means there are roughly as many unfocusable links as there were pseudo-buttons before this PR.

File: templates/admin/user/edit.tmpl (hunk 3 above)
Comment (delvh) [reply templates/admin/user/edit.tmpl]: Ah no, I misremembered something:
If I don't set it, the form will be submitted before you even had the chance to answer the deletion dialogue.
Will add a comment.

File: templates/repo/activity.tmpl (hunk 1 above)
Comment (wxiaoguang) [reply templates/repo/activity.tmpl]: Oh .... it's inside a dropdown .... it shouldn't be a `<button>` 😂

Then I guess the `ui button` classes should be on the `ui dropdown` instead of an inner element, like others?

Either is fine, just curious.

File: templates/base/delete_modal_actions.tmpl (hunk 4 above)
Comment (wxiaoguang) [reply templates/base/delete_modal_actions.tmpl]: I prefer to only use JS , instead of writing `type=button` for every button.

My opinions are:  Gitea UI won't work without JS, the model dialog  won't be shown if the JS doesn't work, so we can choose the simplest&clearest way to write buttons. Even if we write  some "type=button" now, in the future, many developers still won't write it , because  without it the buttons still works well.

File: templates/admin/notice.tmpl (hunk 5 above)
Comment (wxiaoguang) [reply templates/admin/notice.tmpl:45]: If it has been tested, then it's good to keep it. 👍

File: templates/admin/user/edit.tmpl (hunk 3 above)
Comment (wxiaoguang) [reply templates/admin/user/edit.tmpl]: Then the  JS must be fixed .....

File: templates/admin/user/edit.tmpl (hunk 6 above)
Comment (wxiaoguang) [reply templates/admin/user/edit.tmpl]: I believe there was no strong reason. In old code, the AJAX buttons are mostly written as `<a>`, but I think it's more correct to write them as `<button>` now, otherwise there is still a a11y problem .....

File: templates/admin/user/edit.tmpl (hunk 3 above)
Comment (delvh) [reply templates/admin/user/edit.tmpl]: What JS?
At the moment, this is a browser native mechanism:
`<button>`s get an implicit `type="submit"` if none is set.
If I don't set `type="button"`, then it will submit the form as soon as the button is pressed, which otherwise only happens when I click on the `Delete` button inside the deletion dialogue, which is an implicit `submit` button as well.

File: templates/repo/issue/view_content/update_branch_by_merge.tmpl (hunk 7 above)
Comment (wxiaoguang) [reply templates/repo/issue/view_content/update_branch_by_merge.tmpl]: Instead of reverting, `<a>` seems better, then it can be handled by `aria.js`, which only handles Enter for `<a>` or `.js-aria-clickable` to avoid double-enter.

File: templates/admin/user/edit.tmpl (hunk 3 above)
Comment (wxiaoguang) [reply templates/admin/user/edit.tmpl]: Fix `$('.show-modal').on('click', function () {`, add `e.preventDefault()`. Without the preventDefault, indeed there is a JS bug.

File: templates/repo/issue/view_content/update_branch_by_merge.tmpl (hunk 7 above)
Comment (delvh) [reply templates/repo/issue/view_content/update_branch_by_merge.tmpl]: Wait, I'm confused now.
Isn't that exactly the case described in https://github.com/go-gitea/gitea/pull/23337#discussion_r1130842438?

Comment (wxiaoguang) [reply templates/repo/issue/view_content/update_branch_by_merge.tmpl]: No, items in dropdown are different. `aria.js` only  handles `a.item` or `.item.js-aria-clickable` CSS selectors in dropdown.

----

https://github.com/go-gitea/gitea/blob/main/web_src/js/features/aria.js#L81

File: templates/base/delete_modal_actions.tmpl (hunk 4 above)
Comment (delvh) [reply templates/base/delete_modal_actions.tmpl]: Last question:
Do we keep the already existing `type="button"`s in, or should I remove them again?

Comment (wxiaoguang) [reply templates/base/delete_modal_actions.tmpl]: If we decided to use JS and make JS correct, then I think we can remove `type="button"` to keep clear? Either is fine to me.

Comment (delvh) [reply templates/base/delete_modal_actions.tmpl]: Good, then I'm removing the attributes agian now.

File: web_src/js/features/common-global.js [hunk 8]
```diff
@@ -339,7 +339,8 @@ export function initGlobalButtons() {
     alert('Nothing to hide');
   });
 
-  $('.show-modal').on('click', function () {
+  $('.show-modal').on('click', function (e) {
```
Comment (wxiaoguang) [inline web_src/js/features/common-global.js:350]: ps: after reading code, I think all "button" click event handlers in `initGlobalButtons` need to have `preventDefault`, there are 2 more remaining:

* `$('.delete-post.button').on('click', function () {`
* `$('.show-panel.button').on('click', function () {`

File: web_src/js/features/aria.js [hunk 9]
```diff
@@ -96,6 +96,11 @@ function attachOneDropdownAria($dropdown) {
   $dropdown.on('keyup', (e) => { if (e.key.startsWith('Arrow')) deferredRefreshAria(); });
 }
 
+export function initCancelButtons(rootElement=document) {
+  // Only cancel buttons should have the 'cancel' class, and these buttons should not submit an underlying form, so set 'type="button"' for them
+  rootElement.querySelectorAll('.cancel').forEach(cancelButton => cancelButton.setAttribute('type', 'button'));
+}
+
```
Comment (wxiaoguang) [inline web_src/js/features/aria.js]: Do we really need this? 🤔   

I think I fully understand the problem now.

* For Fomantic Dialog:  Maybe Fomantic UI doesn't do `preventDefault` for the `cancel button`, so we need the `type=button` on it if we decided to use `<button>`, or use `<div role=button tabindex=0>` to show that we do not want a form button.
* For other buttons: `initGlobalButtons` would `preventDefault` correctly, so these buttons do not need `type`.

ps: `initCancelButtons` doesn't work for dynamically generated buttons (eg: AJAX?) in the future. So I think we had better avoid it ....

Comment (delvh) [reply web_src/js/features/aria.js]: The good thing is: I can give your words directly back to you now :grin:
We can't ensure that we will forever stay with Fomantic UI (we seem to be on a really slow path to migrating to tailwind in the last few months).
Furthermore, we can't expect future developers to know exactly under which circumstances their buttons will misbehave.
So, with this change, they certainly know that all `cancel` buttons automatically behave correctly, no matter if Fomantic or we do some black magic or not.

> ps: `initCancelButtons` doesn't work for dynamically generated buttons (eg: AJAX?) in the future.

That's why I gave it a `rootElement` parameter. That way, even dynamically generated cancel buttons will be able to be converted.
Furthermore, I'm wondering under what circumstances we need to dynamically generate a cancel button?

File: web_src/js/features/common-global.js [hunk 10]
```diff
@@ -49,20 +49,20 @@ export function initFootLanguageMenu() {
...
-  $(document).on('keypress', '.ui.button', (e) => {
-    if (e.keyCode === 13 || e.keyCode === 32) { // enter key or space bar
-      if (e.target.nodeName === 'BUTTON') return; // button already handles space&enter correctly
-      $(e.target).trigger('click');
-      e.preventDefault();
+  $(document).on('keypress', '.ui.button', (event) => {
+    if (event.keyCode === 13 || event.keyCode === 32) { // enter key or space bar
+      if (event.target.nodeName === 'BUTTON') return; // button already handles space&enter correctly
+      $(event.target).trigger('click');
+      event.preventDefault();
```
Comment (delvh) [inline web_src/js/features/common-global.js]: These changes were simply made for consistency with the rest of the file.

File: web_src/js/features/aria.js (hunk 9 above)
Comment (wxiaoguang) [reply web_src/js/features/aria.js]: > Furthermore, I'm wondering under what circumstances we need to dynamically generate a cancel button?

Partially reloaded page? I can see there will be more and more AJAX requests in the future.

> We can't ensure that we will forever stay with Fomantic UI (we seem to be on a really slow path to migrating to tailwind in the last few months).

I just realized that this problem is only for Fomantic UI, because Gitea misued Fomantic UI modal.  Fomantic UI modal never expects the "ok" button to submit a form, never expects the "cancel" button is a form buton. So it doesn't do "preventDefault" on "cancel button"

Now, Gitea mixes everything together, so developers need to pay attention for the "form button" problem from now on ....

So I think maybe we don't need a "general cancel fix (initCancelButtons)"  to add anything.

> Furthermore, we can't expect future developers to know exactly under which circumstances their buttons will misbehave.

I think just like other `.show-panel.button`, we can do `$.on('.ui.modal .actions .cancel.button') { preventDefault }`, it's general enough if needed.

---

The whole picture (I think it could be the final one in my mind, sorry for the long discussion, it's just for reference & information):

* For our buttons like `.show-panel`, we just add `prevnetDefault`, that's good enough.
* Using "cancel button" in Fomantic UI modal with form needs to prevent the "submit action".
* JS `initCancelButtons` and `$.on('.ui.modal .actions .cancel.button')` all affect performance.

The solutions for "cancel button" in Fomantic UI modal could be either:

1. use `<button type=button>` or `<div role=button tabindex=0>`
2. use `$.on('.ui.modal .actions .cancel.button')`  to catch all events
3. use `initCancelButtons`

I think the first solution could be the most friendly for performance (no affect), it just needs documents and comments (how Gitea misuses the buttons and why the code must be like this ...)

The second solution could be general enough, newly created buttons could also be handled by it.

The third one works, but I prefer the first or the second (a little more for the second, see comment below)

Comment (wxiaoguang) [reply web_src/js/features/aria.js]: One more explanation:

If we do not care about performance too much, the second one is the best and clearest one, and it just works like other button handlers. Pretty consistent.

If we really care about performance, the first one wins, just make it clear for future developers why the code must be written like this.

If you prefer the third one, the selector could be changed to a more accurate one to avoid side effect (like the selector used in the second one).

----

For reference: https://github.com/delvh/gitea/pull/2

File: templates/admin/user/edit.tmpl [hunk 11]
```diff
@@ -213,16 +213,7 @@
...
-				{{.locale.Tr "modal.no"}}
-			</div>
-			<button class="ui green basic inverted ok button">
-				{{svg "octicon-check"}}
-				{{.locale.Tr "modal.yes"}}
-			</button>
-		</div>
-	</form>
+		{{template "base/delete_modal_actions" .}}
+</form>
```
Comment (silverwind) [inline templates/admin/user/edit.tmpl:217]:
```suggestion
  </form>
```

Comment (silverwind) [inline templates/admin/user/edit.tmpl]:
```suggestion
  </form>
```

File: templates/repo/editor/patch.tmpl [hunk 12]
```diff
@@ -45,14 +45,14 @@
...
-				<i class="remove icon"></i>
+			<button class="ui red basic inverted cancel button">
+				{{svg "octicon-x"}}
 				{{.locale.Tr "repo.editor.cancel"}}
-			</div>
-			<div class="ui green basic ok inverted button">
-				<i class="save icon"></i>
+			</button>
+			<button class="ui green basic inverted ok button">
+				<i class="save icon"></i>{{/* TODO: Is there no fitting octicon? */}}
```
Comment (silverwind) [inline templates/repo/editor/patch.tmpl]: Could just port the fontawesome icon:

```svg
<svg xmlns="http://www.w3.org/2000/svg" width="512" height="512" viewBox="0 0 448 512"><path d="m434 130-84-84a48 48 0 0 0-33.9-14H48A48 48 0 0 0 0 80v352a48 48 0 0 0 48 48h352a48 48 0 0 0 48-48V163.9a48 48 0 0 0-14-34zM224 416a64 64 0 1 1 0-128 64 64 0 0 1 0 128zm96-304.5V212a12 12 0 0 1-12 12H76a12 12 0 0 1-12-12V108a12 12 0 0 1 12-12h228.5a12 12 0 0 1 8.5 3.5l3.5 3.5a12 12 0 0 1 3.5 8.5z"/></svg>
```

Save as `web_src/svg/fontawesome-save.svg` and run `make svg`.

File: web_src/js/features/aria.js [hunk 13]
```diff
@@ -96,6 +96,11 @@ function attachOneDropdownAria($dropdown) {
   $dropdown.on('keyup', (e) => { if (e.key.startsWith('Arrow')) deferredRefreshAria(); });
 }
 
+export function initCancelButtons(rootElement=document) {
```
Comment (silverwind) [inline web_src/js/features/aria.js]:
```suggestion
export function initCancelButtons(rootElement = document) {
```

Prefer spaces 😉

File: web_src/js/features/common-global.js (hunk 10 above)
Comment (silverwind) [reply web_src/js/features/common-global.js]: I'd argue `e` is more commonplace across the codebase 😉

**Comment by wxiaoguang** [comment]
According to https://github.com/go-gitea/gitea/blob/main/docs/content/doc/developers/guidelines-refactoring.en-us.md

> If there is no objection from TOC, a refactoring PR could be merged after 7 days with one core member's approval (not the author).

😆

**Review by techknowlogick — APPROVED** [review]

**Comment by techknowlogick** [comment]
> According to https://github.com/go-gitea/gitea/blob/main/docs/content/doc/developers/guidelines-refactoring.en-us.md
> 
> > If there is no objection from TOC, a refactoring PR could be merged after 7 days with one core member's approval (not the author).

Indeed, we have now met that criteria with my +1. I will let someone else be the merger on this, although I suspect with branch rules it'll need to be a TOC member who can use 🧙 powers. Of course, if it doesn't get merged in enough time I can be the merger if needed (please ping if that is the case).

**Review by silverwind — APPROVED** [review]

**Comment by wxiaoguang** [comment]
Does this PR really need to be backported? I am a little worried about whether it would break something, because the button behavior / focus behavior change more or less. Keep it in 1.20 for long enough time before backport?

**Comment by wxiaoguang** [comment]
Caught one 500 (not sure whether it's related)

http://localhost:3000/user/settings/repos

template: base/delete_modal_actions:4:11: executing "base/delete_modal_actions" at <.locale.Tr>: can't evaluate field locale in type string
